pkg/testing/tools/artifacts_api.go

```diff
@@ -161,6 +170,68 @@ func (aac ArtifactAPIClient) GetVersions(ctx context.Context) (list *VersionList
 	return checkResponseAndUnmarshal[VersionList](resp)
 }
 
+// RemoveUnreleasedVersions from the list
+// There is a period of time when a version is already marked as released
+// but not published on the CDN. This happens when we already have build candidates.
+// This function checks if a version marked as released actually has published artifacts.
+// If there are no published artifacts, the version is removed from the list.
+func (aac ArtifactAPIClient) RemoveUnreleasedVersions(ctx context.Context, vList *VersionList) error {
+	suffix, err := testing.GetPackageSuffix(runtime.GOOS, runtime.GOARCH)
+	if err != nil {
+		return fmt.Errorf("failed to generate the artifact suffix: %w", err)
+	}
+
+	results := make([]string, 0, len(vList.Versions))
+
+	for _, versionItem := range vList.Versions {
+		parsedVersion, err := version.ParseVersion(versionItem)
+		if err != nil {
+			return fmt.Errorf("failed to parse version %s: %w", versionItem, err)
+		}
+		// we check only release versions without `-SNAPSHOT`
+		if parsedVersion.Prerelease() != "" {
+			results = append(results, versionItem)
+			continue
+		}
+		url := fmt.Sprintf("%s/elastic-agent-%s-%s", aac.cdnURL, versionItem, suffix)
+		// using method `HEAD` to avoid downloading the file
+		req, err := http.NewRequestWithContext(ctx, http.MethodHead, url, nil)
+		if err != nil {
+			return fmt.Errorf("failed to create an HTTP request to %q: %w", url, err)
+		}
+
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			return fmt.Errorf("failed to request %q: %w", url, err)
+		}
+
+		// we don't read the response. However, we must drain when it's present,
+		// so the connection can be re-used later, see:
+		// https://cs.opensource.google/go/go/+/refs/tags/go1.22.0:src/net/http/response.go;l=62-64
+		_, _ = io.Copy(io.Discard, resp.Body)
+		_ = resp.Body.Close()
+
+		switch resp.StatusCode {
+		case http.StatusNotFound:
+			continue
+		case http.StatusOK:
+			results = append(results, versionItem)
+			continue
+		default:
+			return fmt.Errorf("unexpected status code from %s - %d", url, resp.StatusCode)
+		}
+	}
+
+	// nothing changed
+	if len(vList.Versions) == len(results) {
+		return nil
+	}
```

Comment:
[NIT]
You could remove it, if they're equal, there is no harm in swamping one for the other.

Reply:
As a rule, I don't mutate the function argument unless necessary.